File: pipeline/browser_shell.go
package main

// browser-shell driver — the Go half of Wave 3's chromium thin shell
// (tools/browser-shell.mjs). One long-lived node process; JSON line
// requests in, one JSON line response out. Go owns every judgement.
//
// Errors from the shell surface as Go errors; a shell crash fails the node.

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type browserShell struct {
	cmd    *exec.Cmd
	stdin  io.WriteCloser
	stdout *bufio.Reader
	reqs   int
}

func startBrowserShell() (*browserShell, error) {
	cmd := exec.Command("node", "tools/browser-shell.mjs")
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}
	cmd.Stderr = os.Stderr // playwright diagnostics flow to the node log
	if err := cmd.Start(); err != nil {
		return nil, err
	}
	return &browserShell{cmd: cmd, stdin: stdin, stdout: bufio.NewReader(stdout)}, nil
}

// roundTrip sends req and returns the raw top-level response fields,
// erroring out on a shell-reported "error" field. call and callRawValue
// each decode those fields their own way (map[string]any vs. raw bytes
// that preserve key order).
func (s *browserShell) roundTrip(req map[string]any) (map[string]json.RawMessage, error) {
	s.reqs++
	b, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}
	if _, err := s.stdin.Write(append(b, '\n')); err != nil {
		return nil, fmt.Errorf("browser-shell: write: %w", err)
	}
	line, err := s.stdout.ReadString('\n')
	if err != nil {
		return nil, fmt.Errorf("browser-shell: read (after %d reqs): %w", s.reqs, err)
	}
	var top map[string]json.RawMessage
	if err := json.Unmarshal([]byte(strings.TrimSpace(line)), &top); err != nil {
		return nil, fmt.Errorf("browser-shell: bad response %q: %w", line, err)
	}
	if raw, ok := top["error"]; ok {
		var e string
		json.Unmarshal(raw, &e)
		return nil, fmt.Errorf("browser-shell: %s", e)
	}
	return top, nil
}

func (s *browserShell) call(req map[string]any) (map[string]any, error) {
	top, err := s.roundTrip(req)
	if err != nil {
		return nil, err
	}
	res := make(map[string]any, len(top))
	for k, raw := range top {
		var v any
		if err := json.Unmarshal(raw, &v); err != nil {
			return nil, fmt.Errorf("browser-shell: bad field %q: %w", k, err)
		}
		res[k] = v
	}
	return res, nil
}

func (s *browserShell) callErr(req map[string]any) error {
	_, err := s.call(req)
	return err
}

// callRawValue is call, but returns the "value" field's raw bytes instead of
// decoding through map[string]any — which would lose the browser's own key
// order (e.g. DOM attribute order from a getAttributeNames() walk).
func (s *browserShell) callRawValue(req map[string]any) (json.RawMessage, error) {
	top, err := s.roundTrip(req)
	if err != nil {
		return nil, err
	}
	return top["value"], nil
}

func (s *browserShell) close() {
	s.call(map[string]any{"op": "close"})
	s.stdin.Close()
	s.cmd.Wait()
}

// page is a typed handle over one shell page.
type bpage struct {
	s  *browserShell
	id int
}

func (s *browserShell) launch() error {
	_, err := s.call(map[string]any{"op": "launch"})
	return err
}

func (s *browserShell) newPage(capture bool) (*bpage, error) {
	return s.newPageOrigin(capture, "")
}

func (s *browserShell) newPageOrigin(capture bool, origin string) (*bpage, error) {
	res, err := s.call(map[string]any{"op": "newPage", "capture": capture, "origin": origin})
	if err != nil {
		return nil, err
	}
	id, _ := res["pageId"].(float64)
	return &bpage{s, int(id)}, nil
}

// newPageErrorsOnly: capture pageerror events only (example-fixture pins
// itself to uncaught errors, not the console/requestfailed net).
func (s *browserShell) newPageErrorsOnly() (*bpage, error) {
	res, err := s.call(map[string]any{"op": "newPage", "capture": true, "captureOnly": "pageerror"})
	if err != nil {
		return nil, err
	}
	id, _ := res["pageId"].(float64)
	return &bpage{s, int(id)}, nil
}

func (p *bpage) gotoURL(url string) error {
	_, err := p.s.call(map[string]any{"op": "goto", "pageId": p.id, "url": url})
	return err
}

// evaluate runs expr; for expressions taking an argument, the shell passes
// it as playwright's `arg` (the expression sees it as its single param).
func (p *bpage) evaluate(expr string) (any, error) {
	res, err := p.s.call(map[string]any{"op": "evaluate", "pageId": p.id, "expr": expr})
	if err != nil {
		return nil, err
	}
	return res["value"], nil
}

// evaluateOrdered is evaluate but preserves the result's own JSON key order
// (an object built in the page by literal property assignment, not a Go map).
func (p *bpage) evaluateOrdered(expr string) (json.RawMessage, error) {
	return p.s.callRawValue(map[string]any{"op": "evaluate", "pageId": p.id, "expr": expr})
}

func (p *bpage) evaluateArg(expr string, arg any) (any, error) {
	res, err := p.s.call(map[string]any{"op": "evaluate", "pageId": p.id, "expr": expr, "arg": arg})
	if err != nil {
		return nil, err
	}
	return res["value"], nil
}

// evaluateFn runs a function-shaped expression ("() => {…}") in the page;
// arg becomes the function's single parameter.
func (p *bpage) evaluateFn(expr string) (any, error) {
	return p.evaluateFnArg(expr, nil)
}

func (p *bpage) evaluateFnArg(expr string, arg any) (any, error) {
	res, err := p.s.call(map[string]any{"op": "evaluateFn", "pageId": p.id, "expr": expr, "arg": arg})
	if err != nil {
		return nil, err
	}
	return res["value"], nil
}

func (p *bpage) setContent(html string) error {
	_, err := p.s.call(map[string]any{"op": "setContent", "pageId": p.id, "html": html})
	return err
}

func (p *bpage) addStyleTagPath(path string) error {
	abs, _ := filepath.Abs(path)
	_, err := p.s.call(map[string]any{"op": "addStyleTag", "pageId": p.id, "path": abs})
	return err
}

func (p *bpage) addStyleTag(content string) error {
	_, err := p.s.call(map[string]any{"op": "addStyleTag", "pageId": p.id, "content": content})
	return err
}

func (p *bpage) addScriptTag(content string) error {
	_, err := p.s.call(map[string]any{"op": "addScriptTag", "pageId": p.id, "content": content})
	return err
}

func (p *bpage) focus(selector string, timeoutMS int) error {
	_, err := p.s.call(map[string]any{"op": "focus", "pageId": p.id, "selector": selector, "timeout": timeoutMS})
	return err
}

func (p *bpage) wheel(dx, dy float64) error {
	_, err := p.s.call(map[string]any{"op": "wheel", "pageId": p.id, "dx": dx, "dy": dy})
	return err
}

// routeAbortExternal blocks http(s) subresource loads (initial-render pin).
func (p *bpage) routeAbortExternal() error {
	_, err := p.s.call(map[string]any{"op": "routeAbortExternal", "pageId": p.id})
	return err
}

func (p *bpage) waitForFunction(expr string, timeoutMS int) error {
	_, err := p.s.call(map[string]any{"op": "waitForFunction", "pageId": p.id, "expr": expr, "timeout": timeoutMS})
	return err
}

func (p *bpage) events() ([]string, error) {
	res, err := p.s.call(map[string]any{"op": "events", "pageId": p.id})
	if err != nil {
		return nil, err
	}
	var out []string
	if raw, ok := res["errors"].([]any); ok {
		for _, e := range raw {
			if s, ok := e.(string); ok {
				out = append(out, s)
			}
		}
	}
	return out, nil
}

func (p *bpage) close() {
	p.s.call(map[string]any{"op": "close", "pageId": p.id})
}

// ---- locator primitives (frame = an iframe selector; "" = the page itself) ----

func (p *bpage) locCount(frame, selector string) (int, error) {
	res, err := p.s.call(map[string]any{"op": "locCount", "pageId": p.id, "frame": frame, "selector": selector})
	if err != nil {
		return 0, err
	}
	return int(res["value"].(float64)), nil
}

func (p *bpage) locWait(frame, selector, state string, timeoutMS int) error {
	_, err := p.s.call(map[string]any{"op": "locWait", "pageId": p.id, "frame": frame, "selector": selector, "state": state, "timeout": timeoutMS})
	return err
}

func (p *bpage) locScroll(frame, selector string, index int) error {
	_, err := p.s.call(map[string]any{"op": "locScroll", "pageId": p.id, "frame": frame, "selector": selector, "index": index})
	return err
}

func (p *bpage) locAttr(frame, selector, attr string) (string, bool, error) {
	res, err := p.s.call(map[string]any{"op": "locAttr", "pageId": p.id, "frame": frame, "selector": selector, "attr": attr})
	if err != nil {
		return "", false, err
	}
	v, ok := res["value"].(string)
	return v, ok, nil
}

type bbox struct {
	X, Y, Width, Height float64
}

func (p *bpage) locBox(frame, selector string, index int) (*bbox, error) {
	res, err := p.s.call(map[string]any{"op": "locBox", "pageId": p.id, "frame": frame, "selector": selector, "index": index})
	if err != nil {
		return nil, err
	}
	m, ok := res["value"].(map[string]any)
	if !ok {
		return nil, nil
	}
	return &bbox{m["x"].(float64), m["y"].(float64), m["width"].(float64), m["height"].(float64)}, nil
}

// locEvalAll evaluates expr over every match (playwright evaluateAll).
func (p *bpage) locEvalAll(frame, selector, expr string) (any, error) {
	return p.locEvalAllArg(frame, selector, expr, nil)
}

func (p *bpage) locEvalAllArg(frame, selector, expr string, arg any) (any, error) {
	res, err := p.s.call(map[string]any{"op": "locEvalAll", "pageId": p.id, "frame": frame, "selector": selector, "expr": expr, "arg": arg})
	if err != nil {
		return nil, err
	}
	return res["value"], nil
}

// locClick clicks the nth match; button "right" for context-menu triggers.
func (p *bpage) locClick(frame, selector string, index int, button string) error {
	return p.locClickTimeout(frame, selector, index, button, 15000)
}

// locClickTimeout is locClick with an explicit actionability timeout —
// page.click's playwright default is 30000ms; locator.click's is 5000ms,
// which example-fixture overran twice under 16-way browser load (a dialog
// trigger click), so the shared default here is 15000ms.
func (p *bpage) locClickTimeout(frame, selector string, index int, button string, timeoutMS int) error {
	_, err := p.s.call(map[string]any{"op": "locClick", "pageId": p.id, "frame": frame, "selector": selector, "index": index, "button": button, "timeout": timeoutMS})
	return err
}

// locEval evaluates expr on the nth match only (playwright $eval semantics —
// locEvalAll would touch every match).
func (p *bpage) locEval(frame, selector, expr string, index int) (any, error) {
	res, err := p.s.call(map[string]any{"op": "locEval", "pageId": p.id, "frame": frame, "selector": selector, "expr": expr, "index": index})
	if err != nil {
		return nil, err
	}
	return res["value"], nil
}

func (p *bpage) mouseMove(x, y float64, steps int) error {
	_, err := p.s.call(map[string]any{"op": "mouseMove", "pageId": p.id, "x": x, "y": y, "steps": steps})
	return err
}

func (p *bpage) waitForLoadState(state string, timeoutMS int) error {
	_, err := p.s.call(map[string]any{"op": "waitForLoadState", "pageId": p.id, "state": state, "timeout": timeoutMS})
	return err
}

func (p *bpage) waitForTimeout(ms int) error {
	_, err := p.s.call(map[string]any{"op": "waitForTimeout", "pageId": p.id, "ms": ms})
	return err
}

func (p *bpage) mouseClick(x, y float64) error {
	_, err := p.s.call(map[string]any{"op": "mouseClick", "pageId": p.id, "x": x, "y": y})
	return err
}

// mouseClickButton: button "right" drives context-menu triggers.
func (p *bpage) mouseClickButton(x, y float64, button string) error {
	_, err := p.s.call(map[string]any{"op": "mouseClick", "pageId": p.id, "x": x, "y": y, "button": button})
	return err
}

// driver runs contract-def open/openShadless playwright code against the
// live page.
func (p *bpage) driver(code string) error {
	_, err := p.s.call(map[string]any{"op": "driver", "pageId": p.id, "code": code})
	return err
}

func (p *bpage) keyPress(key string) error {
	_, err := p.s.call(map[string]any{"op": "keyPress", "pageId": p.id, "key": key})
	return err
}

File: pipeline/src/oracle/browser_shell.rs
//! Port of pipeline/browser_shell.go — the Go half of the chromium thin
//! shell (tools/browser-shell.mjs). One long-lived node process; JSON line
//! requests in, one JSON line response out. The driver owns every judgement.
//! Errors from the shell surface as Rust errors; a shell crash fails the node.

use serde_json::{json, Value};
use std::io::{BufRead, BufReader, Write};
use std::process::{Child, ChildStdin, Command, Stdio};
use std::sync::Mutex;

pub struct BrowserShell {
    child: Mutex<Child>,
    stdin: Mutex<Option<ChildStdin>>,
    stdout: Mutex<BufReader<std::process::ChildStdout>>,
    reqs: Mutex<usize>,
}

impl BrowserShell {
    pub fn start() -> Result<BrowserShell, String> {
        let mut cmd = Command::new("node")
            .arg("tools/browser-shell.mjs")
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::inherit()) // playwright diagnostics flow to the node log
            .spawn()
            .map_err(|e| e.to_string())?;
        let stdin = cmd.stdin.take().expect("piped stdin");
        let stdout = cmd.stdout.take().expect("piped stdout");
        Ok(BrowserShell {
            child: Mutex::new(cmd),
            stdin: Mutex::new(Some(stdin)),
            stdout: Mutex::new(BufReader::new(stdout)),
            reqs: Mutex::new(0),
        })
    }

    /// Sends req and returns the raw top-level response, erroring out on a
    /// shell-reported "error" field.
    pub fn round_trip(&self, req: &Value) -> Result<serde_json::Map<String, Value>, String> {
        *self.reqs.lock().unwrap() += 1;
        let n = *self.reqs.lock().unwrap();
        let mut line = serde_json::to_string(req).map_err(|e| e.to_string())?;
        line.push('\n');
        {
            let mut guard = self.stdin.lock().unwrap();
            match guard.as_mut() {
                Some(si) => {
                    si.write_all(line.as_bytes())
                        .map_err(|e| format!("browser-shell: write: {}", e))?;
                    si.flush().ok();
                }
                None => return Err("browser-shell: stdin closed".to_string()),
            }
        }
        let mut resp = String::new();
        {
            let mut so = self.stdout.lock().unwrap();
            so.read_line(&mut resp)
                .map_err(|e| format!("browser-shell: read (after {} reqs): {}", n, e))?;
        }
        let top: serde_json::Map<String, Value> = serde_json::from_str(resp.trim())
            .map_err(|e| format!("browser-shell: bad response {:?}: {}", resp, e))?;
        if let Some(e) = top.get("error") {
            let msg = e.as_str().unwrap_or("unknown").to_string();
            return Err(format!("browser-shell: {}", msg));
        }
        Ok(top)
    }

    pub fn call(&self, req: &Value) -> Result<Value, String> {
        let top = self.round_trip(req)?;
        Ok(Value::Object(top))
    }

    pub fn call_err(&self, req: &Value) -> Result<(), String> {
        self.call(req).map(|_| ())
    }

    /// call, but returns the "value" field's raw bytes instead of decoding
    /// through a map — which would lose the browser's own key order.
    pub fn call_raw_value(&self, req: &Value) -> Result<Value, String> {
        let top = self.round_trip(req)?;
        Ok(top.get("value").cloned().unwrap_or(Value::Null))
    }

    pub fn launch(&self) -> Result<(), String> {
        self.call_err(&json!({"op": "launch"}))
    }

    pub fn new_page(&self, capture: bool) -> Result<BPage<'_>, String> {
        self.new_page_origin(capture, "")
    }

    pub fn new_page_origin(&self, capture: bool, origin: &str) -> Result<BPage<'_>, String> {
        let res = self.call(&json!({"op": "newPage", "capture": capture, "origin": origin}))?;
        let id = res.get("pageId").and_then(|v| v.as_i64()).unwrap_or(0);
        Ok(BPage { s: self, id: id as i32 })
    }

    /// capture pageerror events only (example-fixture pins itself to uncaught
    /// errors, not the console/requestfailed net).
    pub fn new_page_errors_only(&self) -> Result<BPage<'_>, String> {
        let res = self.call(&json!({"op": "newPage", "capture": true, "captureOnly": "pageerror"}))?;
        let id = res.get("pageId").and_then(|v| v.as_i64()).unwrap_or(0);
        Ok(BPage { s: self, id: id as i32 })
    }

    pub fn close(&self) {
        let _ = self.call(&json!({"op": "close"}));
        // drop stdin so the shell sees EOF and exits (Go: stdin.Close())
        let _ = self.stdin.lock().unwrap().take();
        if let Ok(mut child) = self.child.lock() {
            let _ = child.wait();
        }
    }
}

/// A typed handle over one shell page.
pub struct BPage<'a> {
    s: &'a BrowserShell,
    pub id: i32,
}

impl<'a> BPage<'a> {
    pub fn goto_url(&self, url: &str) -> Result<(), String> {
        self.s.call_err(&json!({"op": "goto", "pageId": self.id, "url": url}))
    }

    /// Runs expr, a plain (non-function-shaped) expression, in the page.
    pub fn evaluate(&self, expr: &str) -> Result<Value, String> {
        let res = self
            .s
            .call(&json!({"op": "evaluate", "pageId": self.id, "expr": expr}))?;
        Ok(res.get("value").cloned().unwrap_or(Value::Null))
    }

    /// evaluate but preserves the result's own JSON key order.
    pub fn evaluate_ordered(&self, expr: &str) -> Result<Value, String> {
        self.s
            .call_raw_value(&json!({"op": "evaluate", "pageId": self.id, "expr": expr}))
    }

    /// Runs a function-shaped expression ("() => {…}") in the page; arg
    /// becomes the function's single parameter.
    pub fn evaluate_fn(&self, expr: &str) -> Result<Value, String> {
        self.evaluate_fn_arg(expr, Value::Null)
    }

    pub fn evaluate_fn_arg(&self, expr: &str, arg: Value) -> Result<Value, String> {
        let res = self.s.call(&json!({
            "op": "evaluateFn", "pageId": self.id, "expr": expr, "arg": arg
        }))?;
        Ok(res.get("value").cloned().unwrap_or(Value::Null))
    }

    pub fn set_content(&self, html: &str) -> Result<(), String> {
        self.s
            .call_err(&json!({"op": "setContent", "pageId": self.id, "html": html}))
    }

    pub fn add_style_tag_path(&self, path: &str) -> Result<(), String> {
        let abs = std::fs::canonicalize(path)
            .map(|p| p.to_string_lossy().into_owned())
            .unwrap_or_else(|_| path.to_string());
        self.s
            .call_err(&json!({"op": "addStyleTag", "pageId": self.id, "path": abs}))
    }

    pub fn add_style_tag(&self, content: &str) -> Result<(), String> {
        self.s
            .call_err(&json!({"op": "addStyleTag", "pageId": self.id, "content": content}))
    }

    pub fn add_script_tag(&self, content: &str) -> Result<(), String> {
        self.s
            .call_err(&json!({"op": "addScriptTag", "pageId": self.id, "content": content}))
    }

    pub fn focus(&self, selector: &str, timeout_ms: i64) -> Result<(), String> {
        self.s.call_err(&json!({
            "op": "focus", "pageId": self.id, "selector": selector, "timeout": timeout_ms
        }))
    }

    pub fn wheel(&self, dx: f64, dy: f64) -> Result<(), String> {
        self.s
            .call_err(&json!({"op": "wheel", "pageId": self.id, "dx": dx, "dy": dy}))
    }

    /// Blocks http(s) subresource loads (initial-render pin).
    pub fn route_abort_external(&self) -> Result<(), String> {
        self.s
            .call_err(&json!({"op": "routeAbortExternal", "pageId": self.id}))
    }

    pub fn wait_for_function(&self, expr: &str, timeout_ms: i64) -> Result<(), String> {
        self.s.call_err(&json!({
            "op": "waitForFunction", "pageId": self.id, "expr": expr, "timeout": timeout_ms
        }))
    }

    pub fn events(&self) -> Result<Vec<String>, String> {
        let res = self.s.call(&json!({"op": "events", "pageId": self.id}))?;
        let mut out = Vec::new();
        if let Some(errors) = res.get("errors").and_then(|v| v.as_array()) {
            for e in errors {
                if let Some(s) = e.as_str() {
                    out.push(s.to_string());
                }
            }
        }
        Ok(out)
    }

    pub fn close(&self) {
        let _ = self.s.call(&json!({"op": "close", "pageId": self.id}));
    }

    // ---- locator primitives (frame = an iframe selector; "" = the page) ----

    pub fn loc_count(&self, frame: &str, selector: &str) -> Result<i64, String> {
        let res = self.s.call(&json!({
            "op": "locCount", "pageId": self.id, "frame": frame, "selector": selector
        }))?;
        Ok(res.get("value").and_then(|v| v.as_i64()).unwrap_or(0))
    }

    pub fn loc_wait(&self, frame: &str, selector: &str, state: &str, timeout_ms: i64) -> Result<(), String> {
        self.s.call_err(&json!({
            "op": "locWait", "pageId": self.id, "frame": frame,
            "selector": selector, "state": state, "timeout": timeout_ms
        }))
    }

    pub fn loc_scroll(&self, frame: &str, selector: &str, index: i64) -> Result<(), String> {
        self.s.call_err(&json!({
            "op": "locScroll", "pageId": self.id, "frame": frame,
            "selector": selector, "index": index
        }))
    }

    /// Returns (value, present).
    pub fn loc_attr(&self, frame: &str, selector: &str, attr: &str) -> Result<(String, bool), String> {
        let res = self.s.call(&json!({
            "op": "locAttr", "pageId": self.id, "frame": frame, "selector": selector, "attr": attr
        }))?;
        match res.get("value").and_then(|v| v.as_str()) {
            Some(v) => Ok((v.to_string(), true)),
            None => Ok((String::new(), false)),
        }
    }

    pub fn loc_box(&self, frame: &str, selector: &str, index: i64) -> Result<Option<BBox>, String> {
        let res = self.s.call(&json!({
            "op": "locBox", "pageId": self.id, "frame": frame,
            "selector": selector, "index": index
        }))?;
        let Some(m) = res.get("value").and_then(|v| v.as_object()) else {
            return Ok(None);
        };
        let g = |k: &str| m.get(k).and_then(|v| v.as_f64()).unwrap_or(0.0);
        Ok(Some(BBox { x: g("x"), y: g("y"), width: g("width"), height: g("height") }))
    }

    /// Evaluates expr over every match (playwright evaluateAll).
    pub fn loc_eval_all(&self, frame: &str, selector: &str, expr: &str) -> Result<Value, String> {
        self.loc_eval_all_arg(frame, selector, expr, Value::Null)
    }

    pub fn loc_eval_all_arg(&self, frame: &str, selector: &str, expr: &str, arg: Value) -> Result<Value, String> {
        let res = self.s.call(&json!({
            "op": "locEvalAll", "pageId": self.id, "frame": frame,
            "selector": selector, "expr": expr, "arg": arg
        }))?;
        Ok(res.get("value").cloned().unwrap_or(Value::Null))
    }

    /// Clicks the nth match; button "right" for context-menu triggers.
    pub fn loc_click(&self, frame: &str, selector: &str, index: i64, button: &str) -> Result<(), String> {
        self.loc_click_timeout(frame, selector, index, button, 15000)
    }

    /// locClick with an explicit actionability timeout (15000ms shared
    /// default: example-fixture overran the 5000ms locator default twice
    /// under 16-way browser load).
    pub fn loc_click_timeout(&self, frame: &str, selector: &str, index: i64, button: &str, timeout_ms: i64) -> Result<(), String> {
        self.s.call_err(&json!({
            "op": "locClick", "pageId": self.id, "frame": frame, "selector": selector,
            "index": index, "button": button, "timeout": timeout_ms
        }))
    }

    /// Evaluates expr on the nth match only (playwright $eval semantics).
    pub fn loc_eval(&self, frame: &str, selector: &str, expr: &str, index: i64) -> Result<Value, String> {
        let res = self.s.call(&json!({
            "op": "locEval", "pageId": self.id, "frame": frame,
            "selector": selector, "expr": expr, "index": index
        }))?;
        Ok(res.get("value").cloned().unwrap_or(Value::Null))
    }

    pub fn mouse_move(&self, x: f64, y: f64, steps: i64) -> Result<(), String> {
        self.s.call_err(&json!({
            "op": "mouseMove", "pageId": self.id, "x": x, "y": y, "steps": steps
        }))
    }

    pub fn wait_for_load_state(&self, state: &str, timeout_ms: i64) -> Result<(), String> {
        self.s.call_err(&json!({
            "op": "waitForLoadState", "pageId": self.id, "state": state, "timeout": timeout_ms
        }))
    }

    pub fn wait_for_timeout(&self, ms: i64) -> Result<(), String> {
        self.s
            .call_err(&json!({"op": "waitForTimeout", "pageId": self.id, "ms": ms}))
    }

    pub fn mouse_click(&self, x: f64, y: f64) -> Result<(), String> {
        self.s
            .call_err(&json!({"op": "mouseClick", "pageId": self.id, "x": x, "y": y}))
    }

    /// button "right" drives context-menu triggers.
    pub fn mouse_click_button(&self, x: f64, y: f64, button: &str) -> Result<(), String> {
        self.s.call_err(&json!({
            "op": "mouseClick", "pageId": self.id, "x": x, "y": y, "button": button
        }))
    }

    /// Runs contract-def open/openShadless playwright code against the page.
    pub fn driver(&self, code: &str) -> Result<(), String> {
        self.s
            .call_err(&json!({"op": "driver", "pageId": self.id, "code": code}))
    }

    pub fn key_press(&self, key: &str) -> Result<(), String> {
        self.s
            .call_err(&json!({"op": "keyPress", "pageId": self.id, "key": key}))
    }
}

pub struct BBox {
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
}

impl Drop for BrowserShell {
    fn drop(&mut self) {
        // Best effort: closing stdin trips the driver's EOF handler, which
        // closes chromium — an error path that never reached close() must
        // not leak a browser. Never blocks long on a wedged child.
        if let Some(si) = self.stdin.lock().unwrap().take() {
            drop(si); // dropping the ChildStdin closes the pipe
        }
        let Ok(child) = self.child.get_mut() else {
            return;
        };
        for _ in 0..10 {
            match child.try_wait() {
                Ok(Some(_)) => return,
                Ok(None) => std::thread::sleep(std::time::Duration::from_millis(100)),
                Err(_) => return,
            }
        }
    }
}

